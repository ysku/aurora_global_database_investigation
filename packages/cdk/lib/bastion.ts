import {
  GenericLinuxImage,
  Instance,
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  Vpc,
} from '@aws-cdk/aws-ec2';
import { ManagedPolicy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type BastionProps = {
  dbAccessSG: SecurityGroup;
  vpc: Vpc;
} & StackProps;

export class BastionStack extends Stack {
  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const { dbAccessSG, vpc } = props;

    new Instance(this, 'bastion', {
      instanceName: 'bastion',
      vpc,
      instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO),
      machineImage: new GenericLinuxImage({
        'ap-northeast-1': 'ami-059b6d3840b03d6dd', // Ubuntu Server 20.04 LTS (HVM), SSD Volume Type
      }),
      securityGroup: dbAccessSG,
      role: new Role(this, 'bastion-role', {
        assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [
          ManagedPolicy.fromAwsManagedPolicyName(
            'AmazonSSMManagedInstanceCore'
          ),
        ],
      }),
    });
  }
}
