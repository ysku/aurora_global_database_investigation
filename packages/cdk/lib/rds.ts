import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  SubnetType,
  Vpc,
} from '@aws-cdk/aws-ec2';
import {
  AuroraPostgresEngineVersion,
  Credentials,
  DatabaseCluster,
  DatabaseClusterEngine,
  ParameterGroup,
} from '@aws-cdk/aws-rds';
import { Secret } from '@aws-cdk/aws-secretsmanager';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type RDSStackProps = {
  dbClusterIdentifier: string;
  dbName: string;
  dbSG: SecurityGroup;
  dbUserName: string;
  vpc: Vpc;
} & StackProps;

export class RDSStack extends Stack {
  readonly dbCluster: DatabaseCluster;
  readonly secret: Secret;

  constructor(scope: App, id: string, props: RDSStackProps) {
    super(scope, id, props);

    const { dbClusterIdentifier, dbName, dbSG, dbUserName, vpc } = props;

    // use secret manager to configure database username and password
    // cf. https://docs.aws.amazon.com/ja_jp/secretsmanager/latest/userguide/intro.html
    this.secret = new Secret(this, 'Secret', {
      secretName: `${id}Secret`,
      description: 'RDS database auto-generated user password',
      generateSecretString: {
        excludeCharacters: ':/?#[]@!$&"\'()*+,;=%{}|\\^~[]`',
        generateStringKey: 'password',
        passwordLength: 32,
        secretStringTemplate: `{"username": "${dbUserName}"}`,
      },
    });

    const parameterGroup = new ParameterGroup(this, 'RDSParameterGroup', {
      description: `parameter group for ${dbName}`,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      parameters: {
        log_statement: 'all',
      },
    });

    this.dbCluster = new DatabaseCluster(this, 'RDS', {
      clusterIdentifier: dbClusterIdentifier,
      // cannot use upper case characters.
      defaultDatabaseName: dbName,
      parameterGroup: parameterGroup,
      instances: 2,
      instanceProps: {
        instanceType: InstanceType.of(InstanceClass.R4, InstanceSize.LARGE),
        vpc: vpc,
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED,
        },
        securityGroups: [dbSG],
      },
      port: 5432,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      credentials: Credentials.fromSecret(this.secret),
    });
  }
}
