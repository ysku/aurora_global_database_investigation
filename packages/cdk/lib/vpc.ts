import {
  CfnSecurityGroupIngress,
  SecurityGroup,
  SubnetType,
  Vpc,
  VpcProps,
} from '@aws-cdk/aws-ec2';
import { App, Stack, StackProps } from '@aws-cdk/core';

export class VPCStack extends Stack {
  readonly vpc: Vpc;
  readonly dbAccessSG: SecurityGroup;
  readonly dbSG: SecurityGroup;
  readonly instanceSG: SecurityGroup;

  constructor(scope: App, id: string, props?: StackProps) {
    super(scope, id, props);

    const vpcProps: VpcProps = {
      cidr: '192.168.0.0/16',
      enableDnsHostnames: true,
      enableDnsSupport: true,
      maxAzs: 3,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: SubnetType.PRIVATE,
        },
        {
          cidrMask: 24,
          name: 'isolated',
          subnetType: SubnetType.ISOLATED,
        },
      ],
    };

    // NOTE: default security group will be also created
    this.vpc = new Vpc(this, 'VPC', vpcProps);

    this.dbAccessSG = new SecurityGroup(this, 'dbaccesssg', {
      vpc: this.vpc,
      description: 'for instances that access database',
      securityGroupName: 'Database Access',
    });

    this.dbSG = new SecurityGroup(this, 'dbsg', {
      vpc: this.vpc,
      description: 'for database',
      securityGroupName: 'Database',
    });

    this.instanceSG = new SecurityGroup(this, 'instancesg', {
      vpc: this.vpc,
      description: 'for ec2 instance',
      securityGroupName: 'Instance',
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'fromdbaccesssg', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.dbAccessSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.dbAccessSG.securityGroupId,
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'frominstance', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.instanceSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.instanceSG.securityGroupId,
    });
  }
}
