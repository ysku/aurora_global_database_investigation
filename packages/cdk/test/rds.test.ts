import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';

import { RDSStack } from '../lib/rds';
import { VPCStack } from '../lib/vpc';

describe('RDSStack', () => {
  const app = new App();
  const vpc = new VPCStack(app, 'TestVPCStack');

  test('snapshot', () => {
    const stack = new RDSStack(app, 'TestRDSStack', {
      dbClusterIdentifier: 'test-db-cluster-name',
      dbName: 'test-db-name',
      dbSG: vpc.dbSG,
      dbUserName: 'test-user',
      vpc: vpc.vpc,
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
