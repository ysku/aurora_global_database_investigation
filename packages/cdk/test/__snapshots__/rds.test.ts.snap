// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RDSStack snapshot 1`] = `
Object {
  "Resources": Object {
    "RDSE0E96D00": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "DBClusterIdentifier": "test-db-cluster-name",
        "DBClusterParameterGroupName": Object {
          "Ref": "RDSParameterGroup973BA941",
        },
        "DBSubnetGroupName": Object {
          "Ref": "RDSSubnets873FC54A",
        },
        "DatabaseName": "test-db-name",
        "Engine": "aurora-postgresql",
        "EngineVersion": "11.9",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "SecretA720EF05",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "SecretA720EF05",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "Port": 5432,
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::ImportValue": "TestVPCStack:ExportsOutputFnGetAttdbsgB988D619GroupId34E3B055",
          },
        ],
      },
      "Type": "AWS::RDS::DBCluster",
      "UpdateReplacePolicy": "Snapshot",
    },
    "RDSInstance173AB26B0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "RDSE0E96D00",
        },
        "DBInstanceClass": "db.r4.large",
        "DBInstanceIdentifier": "test-db-cluster-nameinstance1",
        "DBSubnetGroupName": Object {
          "Ref": "RDSSubnets873FC54A",
        },
        "Engine": "aurora-postgresql",
        "EngineVersion": "11.9",
        "PubliclyAccessible": false,
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "RDSInstance21447102B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "RDSE0E96D00",
        },
        "DBInstanceClass": "db.r4.large",
        "DBInstanceIdentifier": "test-db-cluster-nameinstance2",
        "DBSubnetGroupName": Object {
          "Ref": "RDSSubnets873FC54A",
        },
        "Engine": "aurora-postgresql",
        "EngineVersion": "11.9",
        "PubliclyAccessible": false,
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "RDSParameterGroup973BA941": Object {
      "Properties": Object {
        "Description": "parameter group for test-db-name",
        "Family": "aurora-postgresql11",
        "Parameters": Object {
          "log_statement": "all",
        },
      },
      "Type": "AWS::RDS::DBClusterParameterGroup",
    },
    "RDSSubnets873FC54A": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnets for RDS database",
        "SubnetIds": Array [
          Object {
            "Fn::ImportValue": "TestVPCStack:ExportsOutputRefVPCisolatedSubnet1Subnet84AF70385D9D8D66",
          },
          Object {
            "Fn::ImportValue": "TestVPCStack:ExportsOutputRefVPCisolatedSubnet2Subnet8AE9E7657AFFAE94",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "SecretA720EF05": Object {
      "Properties": Object {
        "Description": "RDS database auto-generated user password",
        "GenerateSecretString": Object {
          "ExcludeCharacters": ":/?#[]@!$&\\"'()*+,;=%{}|\\\\^~[]\`",
          "GenerateStringKey": "password",
          "PasswordLength": 32,
          "SecretStringTemplate": "{\\"username\\": \\"test-user\\"}",
        },
        "Name": "TestRDSStackSecret",
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "SecretAttachment2E1B7C3B": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "SecretA720EF05",
        },
        "TargetId": Object {
          "Ref": "RDSE0E96D00",
        },
        "TargetType": "AWS::RDS::DBCluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
  },
}
`;
